cirq/sim/wave_function.py

```diff
@@ -282,3 +286,40 @@ def _validate_indices(num_qubits: int, indices: List[int]) -> None:
     if any(index >= num_qubits for index in indices):
         raise IndexError('Out of range indices, must be less than number of '
                          'qubits but was {}'.format(indices))
+
+
+def sample_terminal_measurements(
```

Comment:
Is there a reason that we are exposing "terminal only" as a constraint on the method, instead of simply picking the appropriate method for the user based on the circuit they gave to us? Just as a way of communicating that this particular case will be much faster?

Reply:
After looking more carefully I realize this method was not doing at all what I expected. We need a better name for it.

Maybe it should be an instance method of `StepResult`?